### PR TITLE
Port new_empty to ATen.

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25583 Port new_full to ATen.
* **#25475 Port new_empty to ATen.**

I got sucked into this rabbit hole when I was trying to understand
what I should do with TensorTypeId occurrences in
torch/csrc/utils/tensor_new.cpp.  I eventually concluded that all of my problems
were because Tensor.new_empty was hand implemented and not actually a native
function.  So I made it a native function.

There are a bunch of other new_* functions which should get this
treatment, but I'm sending out this PR just to show how it can
be done.

The general recipe:
1. Implement a concept of TensorOptions merging (TensorOptions::merge_in).
   This represents the notion of taking a tensor, but "overriding" some
   of its values with specific overrides.  One subtlety here is how
   devices get merged; in this version of the patch, devices override
   just like everything else, and new_* functions are expected to
   setup an appropriate device guard.
2. Implement new_empty as a native function, using options merging.
3. Add another special case to Python binding generation to treat new_*
   similar to *_like (i.e., handle TensorOptions correctly).
4. Since *_new is the first instance of methods with TensorOptions,
   I needed to adjust the binding code to support this.
5. Delete the now defunct manual binding code.
6. Delete manual type annotations that are no longer necessary since
   we're going through native.

I didn't handle memory format correctly here.  I don't know if this function
should accept memory format; prior memory format patches didn't add support
for memory format to new_like.  If we had put memory format in TensorOptions
this wouldn't have been a question.

Differential Revision: [D17133000](https://our.internmc.facebook.com/intern/diff/D17133000/)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25475

Differential Revision: [D17133000](https://our.internmc.facebook.com/intern/diff/D17133000)

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
Overall looks good, but `AttributeError: type object 'torch._C._TensorBase' has no attribute 'new_empty'`

**Comment by gchanan** [comment]
@izdeby this is probably going to affect your TensorOptions refactoring.

**Comment by ezyang** [comment]
@izdeby This PR is non-blocking for other work and I can pick it up after you land TensorOptions.

File: tools/autograd/gen_python_functions.py
Comment (gchanan) [inline tools/autograd/gen_python_functions.py]: just rename the variable to is_like_or_new_function ?

**Review by gchanan — COMMENTED** [review]
does this end up getting generated in variable_factories?

File: tools/autograd/gen_python_functions.py
Comment (ezyang) [reply tools/autograd/gen_python_functions.py]: I did... but then I have to rename more variables XD

Comment (ezyang) [reply tools/autograd/gen_python_functions.py]: (I'll renaming)

**Comment by ezyang** [comment]
@izdeby But I would appreciate knowing if it does conflict or not so I know if to wait or not :)

File: c10/core/TensorOptions.h
Comment (gchanan) [inline c10/core/TensorOptions.h]: I don't really buy canonicalizing this behavior in TensorOptions.

1) Can't you just get rid of this by not having `device_guard: False`
2) If we want this behavior, we can limit it to tensor_new.  I would NOT expect this is what merge_in does.

**Review by gchanan — CHANGES_REQUESTED** [review]

File: c10/core/TensorOptions.h
Comment (ezyang) [reply c10/core/TensorOptions.h]: I felt dirty editing options. But yes, it's not necessary XD

File: c10/core/TensorOptions.h
Comment (ezyang) [reply c10/core/TensorOptions.h]: I think you're right!

File: c10/core/TensorOptions.h
Comment (ezyang) [reply c10/core/TensorOptions.h]: Yeah I'm not sure either. Let CI decide?

**Comment by ezyang** [comment]
OK, every handled except the pass-by-name-and-then-copy.

**Comment by izdeby** [comment]
> @izdeby But I would appreciate knowing if it does conflict or not so I know if to wait or not :)

@ezyang, dont wait for me, thanks

**Comment by ezyang** [comment]
Somehow, the Windows failure is legit:

```
C:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/tmp_bin/nvcc C:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/aten/src/THCUNN/L1Cost.cu -c -o C:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/caffe2/CMakeFiles/torch.dir/__/aten/src/THCUNN/./torch_generated_L1Cost.cu.obj -ccbin "C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.22.27905/bin/Hostx64/x64/cl.exe" -m64 -Dtorch_EXPORTS -D_THP_CORE -DUSE_CUDA -DCAFFE2_BUILD_MAIN_LIB -DAT_PARALLEL_OPENMP=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_c2 -D_CRT_SECURE_NO_DEPRECATE=1 -DWIN32_LEAN_AND_MEAN -DCPUINFO_SUPPORTED_PLATFORM=1 -Xcompiler ,\"/DWIN32\",\"/D_WINDOWS\",\"/GR\",\"/w\",\"/MP\",\"/bigobj\",\"-openmp:experimental\",\"-DHAVE_AVX_CPU_DEFINITION\",\"-DHAVE_AVX2_CPU_DEFINITION\",\"/MD\",\"/O2\",\"/Ob2\",\"/w\",\"/MP\",\"/bigobj\" -Xcompiler /w -w -DONNX_NAMESPACE=onnx_c2 -gencode arch=compute_50,code=sm_50 -Xcudafe --diag_suppress=cc_clobber_ignored -Xcudafe --diag_suppress=integer_sign_change -Xcudafe --diag_suppress=useless_using_declaration -Xcudafe --diag_suppress=set_but_not_used -Xcompiler -MD --expt-relaxed-constexpr --expt-extended-lambda -Xcompiler /wd4819 -Xcompiler /wd4503 -Xcompiler /wd4190 -Xcompiler /wd4244 -Xcompiler /wd4251 -Xcompiler /wd4275 -Xcompiler /wd4522 -Wno-deprecated-gpu-targets --expt-extended-lambda -gencode arch=compute_50,code=sm_50 -DCUDA_HAS_FP16=1 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ -DNVCC "-IC:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include" -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/aten/src -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/aten/src -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/cmake/../third_party/googletest/googlemock/include -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/cmake/../third_party/googletest/googletest/include -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/third_party/protobuf/src -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/cmake/../third_party/benchmark/include -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/cmake/../third_party/eigen -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/cmake/../third_party/pybind11/include -I/opt/rocm/hip/include -I/include -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/cmake/../third_party/cub -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/caffe2/contrib/aten -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/third_party/onnx -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/third_party/onnx -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/third_party/foxi -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/third_party/foxi -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/caffe2/../torch/csrc/api -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/caffe2/../torch/csrc/api/include "-IC:/Program Files/NVIDIA Corporation/NvToolsExt/include" -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/caffe2/aten/src/TH -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/caffe2/aten/src/TH -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/caffe2/../torch/../aten/src -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/caffe2/aten/src -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/caffe2/../aten/src -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/caffe2/../aten/src/ATen -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/caffe2/../torch/csrc -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/caffe2/../torch/../third_party/miniz-2.0.8 -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/aten/src/TH -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/aten/../third_party/catch/single_include -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/aten/src/ATen/.. -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/caffe2/aten/src/ATen -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/third_party/miniz-2.0.8 -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/caffe2/core/nomnigraph/include -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/caffe2/ -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/caffe2/aten/src/THC -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/aten/src/THC -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/aten/src/THCUNN -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/aten/src/ATen/cuda -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/c10/../ -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/third_party/cpuinfo/include -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/third_party/FP16/include -IC:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/c10/cuda/../..
14:06:51 L1Cost.cu
14:06:51 C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/allocator/allocator_traits.inl(101): error C2993: 'T': illegal type for non-type template parameter '__formal'
14:06:51 C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/allocator/allocator_traits.inl(101): note: see reference to class template instantiation 'thrust::detail::allocator_traits_detail::has_member_construct2_impl_has_member<T,Result(Arg1,Arg2)>' being compiled
14:06:51 C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/allocator/allocator_traits.inl(101): error C2065: 't': undeclared identifier
14:06:51 C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/allocator/allocator_traits.inl(101): error C2923: 'std::_Select<__formal>::_Apply': 't' is not a valid template type argument for parameter '<unnamed-symbol>'
14:06:51 C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/allocator/allocator_traits.inl(101): error C2062: type 'unknown-type' unexpected
14:06:51 -- Removing C:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/caffe2/CMakeFiles/torch.dir/__/aten/src/THCUNN/./torch_generated_L1Cost.cu.obj
14:06:51 "C:/Program Files/CMake/bin/cmake.exe" -E remove C:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/caffe2/CMakeFiles/torch.dir/__/aten/src/THCUNN/./torch_generated_L1Cost.cu.obj
14:06:51 CMake Error at torch_generated_L1Cost.cu.obj.Release.cmake:281 (message):
14:06:51   Error generating file
14:06:51   C:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/caffe2/CMakeFiles/torch.dir/__/aten/src/THCUNN/./torch_generated_L1Cost.cu.obj
14:06:51 
```

File: c10/core/Device.h
Comment (gchanan) [inline c10/core/Device.h]: you don't need this anymore?

Comment (ezyang) [reply c10/core/Device.h]: Not for this patch, but it seems like it might be useful for someone else, sometime?

**Comment by ezyang** [comment]
I think the Windows failure is fake. Evidence: https://github.com/pytorch/pytorch/pull/25583 is all green on Windows.

File: c10/core/Device.h
Comment (gchanan) [reply c10/core/Device.h]: nope, I don't think it's useful.  The comment and analysis is all about not setting the device_guard properly.

**Review by gchanan — APPROVED** [review]
approved if you get rid of Device merge_in.

**Pull request closed without merging**
